src/zip.jl

```diff
@@ -0,0 +1,79 @@
+
+struct DiskZip{Is<:Tuple}
+    is::Is
+end
+Base.iterate(dz::DiskZip) = Base.iterate(Iterators.Zip(dz.is))
+Base.iterate(dz::DiskZip, i) = Base.iterate(Iterators.Zip(dz.is), i)
+Base.first(dz::DiskZip) = Base.first(Iterators.Zip(dz.is))
+Base.last(dz::DiskZip) = Base.last(Iterators.Zip(dz.is))
+Base.length(dz::DiskZip) = Base.length(Iterators.Zip(dz.is))
+Base.size(dz::DiskZip) = Base.size(Iterators.Zip(dz.is))
+Base.IteratorSize(::Type{DiskZip{Is}}) where {Is<:Tuple} =
+    Base.IteratorSize(Iterators.Zip{Is})
+Base.IteratorEltype(::Type{DiskZip{Is}}) where {Is<:Tuple} =
+    Base.IteratorEltype(Iterators.Zip{Is})
+
+# Rechunk using the chunks of the first Chunked array
+# This forces the iteration order to be the same for
+# all arrays.
+
+function DiskZip(As::AbstractArray{<:Any,N}...) where N
+    map(As) do A
+        size(A) == size(first(As)) || throw(DimensionMismatch("Arrays zipped with disk arrays must be the same size"))
+    end
+    # Get the chunkes of the first Chunked array
+    chunks = reduce(As; init=nothing) do acc, A
+        if isnothing(acc) && (haschunks(A) isa Chunked)
+            eachchunk(A)
+        else
+            acc
+        end
+    end
+    if isnothing(chunks)
+        return DiskZip(As)
+    else
+        rechunked = map(As) do A
+            RechunkedDiskArray(A, chunks)
+        end
+        return DiskZip(rechunked)
+    end
+end
+# For now we only allow zip on exact same-sized arrays
+DiskZip(As...) = throw(ArgumentError("zip on disk arrays only works with other same-sized AbstractArray"))
+
+# Collect zipped disk arrays in the right order
+function Base.collect(dz::DiskZip)
+    out = similar(first(dz.is), eltype(dz))
+    i = iterate(dz)
+    for I in eachindex(first(dz.is))
```

Comment:
Would it be possbile to do a block-wise copying here instead? I think we would need some nested iterator here, similar to what I defined in Iterators.jl. Maybe we can rather fix that one and adapt, because this implementation here will have plenty of random accesses and will be horribly slow.

Reply:
Turns out its fine as is, the chunking just propagates through.

Reply:
Its amazing actually it seems all the iterators just composes properly though `DiskZip`, `Zip` and each individual stateful iterator. Without doing anything special here.